homeassistant/components/keba/__init__.py

```diff
@@ -0,0 +1,212 @@
+"""Support for KEBA charging stations."""
+import asyncio
+import logging
+
+from keba_kecontact.connection import KebaKeContact
+import voluptuous as vol
+
+from homeassistant.const import CONF_HOST
+from homeassistant.helpers import discovery
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'keba'
+SUPPORTED_COMPONENTS = ['binary_sensor', 'sensor', "lock"]
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional('rfid', default='00845500'): cv.string,
+        vol.Optional('failsafe', default=False): cv.boolean,
+        vol.Optional('failsafe_timeout', default=30): cv.positive_int,
+        vol.Optional('failsafe_fallback', default=6): cv.positive_int,
+        vol.Optional('failsafe_persist', default=0): cv.positive_int,
+        vol.Optional('refresh_interval', default=5): cv.positive_int,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+_SERVICE_MAP = {
+    'request_data': 'request_data',
+    'set_energy': 'async_set_energy',
+    'set_current': 'async_set_current',
+    'authorize': 'async_start',
+    'deauthorize': 'async_stop',
+    'enable': 'async_enable_ev',
+    'disable': 'async_disable_ev',
+    'set_failsafe': 'async_set_failsafe'
+}
+
+
+async def async_setup(hass, config):
+    """Check connectivity and version of KEBA charging station."""
+    host = config[DOMAIN][CONF_HOST]
+    rfid = config[DOMAIN]['rfid']
+    refresh_interval = config[DOMAIN]['refresh_interval']
+    keba = KebaHandler(hass, host, rfid, refresh_interval)
+    hass.data[DOMAIN] = keba
+
+    # Wait for KebaHandler setup complete (initial values loaded)
+    await keba.setup()
+
+    # Set failsafe mode at start up of home assistant
+    failsafe = config[DOMAIN]['failsafe']
+    timeout = config[DOMAIN]['failsafe_timeout'] if failsafe else 0
+    fallback = config[DOMAIN]['failsafe_fallback'] if failsafe else 0
+    persist = config[DOMAIN]['failsafe_persist'] if failsafe else 0
+    try:
+        hass.loop.create_task(keba.set_failsafe(timeout, fallback, persist))
+    except ValueError as ex:
+        _LOGGER.warning("Could not set failsafe mode %s", ex)
+
+    # Register services to hass
+    async def async_execute_service(call):
```

Comment:
If we don't need to await inside the service handler, we can make it a callback by decorating it with `@callback` imported from core.py. Change `async def` to `def`.

Reply:
It is not possible to await the function_call as suggested in the next comment and making it a callback decorated function. Thus I have not used the callback decorater and await the function_call inside.